FEAT #41

SchoofsEbert opened the pull request:
This PR addresses issue #41.  
  
Commit message on merge should read: FEAT #41  
  
This PR makes it possible to set the minimum confidence level for each user, and only add categories if they are above this minimum confidence level.
It also disables the creation of attribute tags at the moment, because in my opinion they really are not helpful. This could be undone if they are deemed meaningful.  

PR should be merged together with https://github.com/LibrePhotos/librephotos-frontend/pull/5.

SchoofsEbert committed:
Only select scene categories with a confidence higher as 0.1

SchoofsEbert committed:
Pushed to many lines

SchoofsEbert committed:
Pushed to many lines

SchoofsEbert committed:
Merge remote-tracking branch 'origin/scene_confidence' into scene_confidence

# Conflicts:
#	api/places365/places365.py

derneuere commented:
Hey SchoofsEbert,

thank you for your pull request. I tried it and it works now way better 👍 

I think the confidence should be configurable in the settings page for the admin, because the confidence value isn't related to the deployment of the docker images. 

I don't think most people would understand what the confidence value does. We should probably test it out to find values that would be equivalent to "Very high accuracy" "High accuracy", "Medium accuracy" and "Low accuracy".

I looked into "indoor" and "outdoor". I don't see it when opening a picture, but I can search for it. If I understand the code correctly, I think it is saved in "self.captions_json", but it isn't added to an album place in https://github.com/LibrePhotos/librephotos/blob/14d2531dadb2b2750af26f80041efeb957b31100/api/models.py#L496-L515

The attributes and the environment should only be added if it reached the confidence value, but it should be independently determined of the categories. Do the attributes and the environment have their own confidence value?

It would be great if you could write some comments explaining what inference_places365() returns and how the returned type is structured. I think you are the only person who knows roughly how it works.

SchoofsEbert committed:
Added some more comments

derneuere commented on api/places365/places365.py:
```diff
@@ -66,14 +66,18 @@ def remove_nonspace_separators(text):
            features_blobs.append(np.squeeze(output.data.cpu().numpy()))
 
-        def load_model():
+        def load_model():  # TODO Should the model be reloaded for every photo? Wouldn't it be better to do that once?
```
No it should not. Good catch 👍

derneuere commented on api/places365/places365.py:
```diff
@@ -117,22 +122,34 @@ def load_model():
+        # Take the dot product of out W_attribute model and the feature blobs
+        # And sort it along the -1 axis
+        # This results in idx_a, with the last elements the index numbers of attributes, we have the most confidence in
+        # Can't seem to get any confidence values, also all the attributes it detect are not really meaningful i.m.o.
```
I think the original author copy and pasted part of this file here: https://github.com/CSAILVision/places365/blob/master/demo_pytorch_CAM.py
It seems to output a .json file. Maybe it contains some probability value.
 
While it is nice to have general attributes, it isn't as nice as object detection.

SchoofsEbert committed:
Make confidence adaptable from setting

derneuere committed:
[REF] Allow attributes to be added

While I understand that the attributes are not that helpful, I worry about error messages from end users because attributes are missing. We should replace this in the future with a better algorithm, instead of not adding any attributes.

derneuere committed:
[REF] Allow attributes for now
